File: app/src/main/java/com/vladislav/workoutassistant/view/fragments/TimePickerFragment.java
package com.vladislav.workoutassistant.view.fragments;

import android.app.Activity;
import android.app.Dialog;
import android.app.TimePickerDialog;
import android.content.Intent;
import android.os.Bundle;
import android.text.format.DateFormat;
import android.widget.TimePicker;

import java.util.Calendar;
import java.util.Date;

import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.Fragment;

public class TimePickerFragment extends DialogFragment {

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        Date date = (Date) getArguments().getSerializable(ARG_TIME);
        Calendar calendar = Calendar.getInstance();
        calendar.setTime(date);
        int hour = calendar.get(Calendar.HOUR_OF_DAY);
        int minute = calendar.get(Calendar.MINUTE);

        return new TimePickerDialog(getActivity(),
                new TimePickerDialog.OnTimeSetListener() {
                    @Override
                    public void onTimeSet(TimePicker view, int hourOfDay, int minute) {
                        Calendar calendar = Calendar.getInstance();
                        calendar.set(Calendar.HOUR, hourOfDay);
                        calendar.set(Calendar.MINUTE, minute);
                        sendResult(Activity.RESULT_OK, calendar.getTime());
                    }
                },
                hour, minute, DateFormat.is24HourFormat(getActivity()));
    }

    private void sendResult(int resultCode, Date date) {
        Fragment targetFragment = getTargetFragment();
        if (targetFragment != null) {
            Intent intent = new Intent();
            intent.putExtra(EXTRA_TIME, date);
            targetFragment.onActivityResult(getTargetRequestCode(), resultCode, intent);
        }
    }

    public static TimePickerFragment newInstance(Date date) {
        Bundle args = new Bundle();
        args.putSerializable(ARG_TIME, date);

        TimePickerFragment fragment = new TimePickerFragment();
        fragment.setArguments(args);

        return fragment;
    }

    private static final String ARG_TIME = "arg_time";
    public static final String EXTRA_TIME = "extra_time";
}

File: common/src/main/java/com/vladislavmyasnikov/common/view/TimePickerFragment.kt
package com.vladislavmyasnikov.common.view

import android.app.Activity
import android.app.Dialog
import android.app.TimePickerDialog
import android.content.Intent
import android.os.Bundle
import android.text.format.DateFormat
import androidx.fragment.app.DialogFragment
import com.vladislavmyasnikov.common.models.TimePoint

class TimePickerFragment : DialogFragment() {

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val initialTime = arguments!!.getParcelable<TimePoint>(TIME_ARG)!!

        return TimePickerDialog(
                activity,
                TimePickerDialog.OnTimeSetListener { _, hourOfDay, minute ->
                    sendResult(Activity.RESULT_OK, TimePoint(hourOfDay.toLong(), minute.toLong()))
                },
                initialTime.hours.toInt(),
                initialTime.minutes.toInt(),
                DateFormat.is24HourFormat(activity)
        )
    }

    private fun sendResult(resultCode: Int, time: TimePoint) {
        targetFragment?.let {
            val intent = Intent().apply {
                putExtra(TIME_EXTRA, time)
            }
            it.onActivityResult(targetRequestCode, resultCode, intent)
        }
    }

    companion object {
        private const val TIME_EXTRA = "time_extra"
        private const val TIME_ARG = "time_arg"

        fun newInstance(time: TimePoint): TimePickerFragment {
            return TimePickerFragment().apply {
                arguments = Bundle().apply {
                    putParcelable(TIME_ARG, time)
                }
            }
        }

        fun extractTime(data: Intent): TimePoint {
            return data.getParcelableExtra(TIME_EXTRA)!!
        }
    }
}
